### PR TITLE
Implement SimKO to add entropy in TopK token sampling during RL

### PR DESCRIPTION
**Pull request by ighoshsubho**
This Impl is directly inspired from this paper - https://arxiv.org/pdf/2510.14807v1

I'm still actively testing out and will share results on it, the topK sampling metrics during training are yet to get logged in W&B

### CODE REVIEW

File: torchtitan/grpo/grpo_step.py [hunk 1]
```diff
@@ -175,6 +175,91 @@ def compose_grpo_loss(
 
     return loss
 
+def compute_token_entropy(pred, mask):
```
Comment (dmahan93) [inline torchtitan/grpo/grpo_step.py:178]: we have a vocab parallel entropy fn in utils, that should be used instead

File: torchtitan/grpo/grpo_step.py [hunk 2]
```diff
@@ -175,6 +175,91 @@ def compose_grpo_loss(
...
+    log_probs = torch.nn.functional.log_softmax(pred, dim=-1)
+    probs = torch.exp(log_probs)
+    entropy = -(probs * log_probs).sum(dim=-1)
+    return entropy * mask
+
+
+def apply_simko_adjustment(
+    ratio, pred, labels, reward, mask,
+    alpha=0.01, K=3, lambda_top1=1.1, tau_percentile=80
+):
```
Comment (dmahan93) [inline torchtitan/grpo/grpo_step.py:198]: this fn does not work with tensor parallel

File: torchtitan/grpo/grpo_step.py [hunk 3]
```diff
@@ -175,6 +175,91 @@ def compose_grpo_loss(
...
+        mask: Attention mask [batch_size, seq_len]
+        alpha: Label smoothing strength
+        K: Number of top candidates
+        lambda_top1: Penalty multiplier for top-1 negatives
+        tau_percentile: Entropy threshold percentile
+    
+    Returns:
+        adjusted_ratio: Ratio with SimKO adjustments
+    """
+    # 1. Identify forking tokens (high-entropy)
```
Comment (dmahan93) [inline torchtitan/grpo/grpo_step.py:216]: does this need backpropagated through, or can you wrap it with no_grad?

File: torchtitan/grpo/grpo_step.py [hunk 4]
```diff
@@ -175,6 +175,91 @@ def compose_grpo_loss(
...
+    valid_entropy = entropy[mask.bool()]
+    if len(valid_entropy) > 0:
+        tau = torch.quantile(valid_entropy, tau_percentile / 100.0)
+        w = (entropy > tau).float()
+    else:
+        w = torch.zeros_like(mask)
+    
+    # 2. Get top-K tokens and compute their ratios
+    _, topk_indices = torch.topk(pred, k=K, dim=-1)
+    new_log_probs_full = torch.nn.functional.log_softmax(pred, dim=-1)
```
Comment (dmahan93) [inline torchtitan/grpo/grpo_step.py:227]: this uses up a lot of memory

**Review by dmahan93 — CHANGES_REQUESTED** [review]

File: torchtitan/grpo/grpo_step.py (hunk 1 above)
Comment (ighoshsubho) [reply torchtitan/grpo/grpo_step.py:178]: sure will do in the following commit.

File: torchtitan/grpo/grpo_step.py (hunk 4 above)
Comment (ighoshsubho) [reply torchtitan/grpo/grpo_step.py:227]: sure, do you think like doing log softmax one time would do? If we are doing a no grad on this, it could be reused for old poilicy top k probs as well

File: torchtitan/grpo/grpo_step.py (hunk 2 above)
Comment (ighoshsubho) [reply torchtitan/grpo/grpo_step.py:198]: ok fixing it

File: torchtitan/grpo/grpo_step.py (hunk 3 above)
Comment (ighoshsubho) [reply torchtitan/grpo/grpo_step.py:216]: No this won't need a backprop and not seen anywhere in paper to mention so, yeah no grad is way to go

File: torchtitan/grpo/grpo_step.py (hunk 4 above)
Comment (dmahan93) [reply torchtitan/grpo/grpo_step.py:227]: well this needs grads, since the grads are flowing through this softmax, we need to combine the way we're currently getting logprobs with the top-k logprob method, as the logprobs we need for the GRPO/GSPO loss are not guaranteed to be within the top-k. the quick and dirty way may be to wrap it in a checkpoint and chunk it, that way we "save" the memory by having it re-execute on backward in manageable chunks, even if it's not optimal, it should still just be a tiny portion of the compute so we can add <figuring out a way to reuse the whole logit space> to the backlog as i'm mostly worried about memory 😎
